### ISSUE TITLE
Upstreaming OpenShift features

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#10408]
This is a tracking issue for the purposes of identifying the pieces of OpenShift we would like to componentize and make more consumable for other users of Kube post 1.0.  This list should record whether we want the component in core Kube, contrib Kube, an ecosystem plugin, or to remain in OpenShift (which is a distribution of Kube with enhancements).  This issue does not track the actual work required to compile/build/include/run these components, just the high level discussion.
### Upstreamed features in progress or completed
- **implemented in deployments** A declarative description of how to iteratively roll out deployments (covered by #1743)
- **implemented in ingresses** a declarative route resource which maps a service into one or more proxies - offers programmable http / HTTPS endpoints and configuration of L7 rules
- **implemented in kubectl create** secret creation and generation from the CLI
- **implemented in kubectl edit** open a CLI editor to edit resources
- **implemented in kubectl export** export resources from the CLI for reuse elsewhere
- **implemented in kubectl drain** evacuate a node
- **in progress** Parameterized config files for simple instantiation of location independent resources
- **in progress** The authorization engine - which can apply policy decisions on Kubernetes resources in an ABAC and RBAC model and manages a policy, roles, role bindings
- **pending authorization** The OAuth daemon (which manages OAuth clients, tokens, and apps and can authenticate bearer tokens to the API)
- **pending authorization** The user / identity API which allows third party auth integrations to track users by user name and store them in etcd, as well as group integration for folks integrating with LDAP
- **pending** Move/unify the HAProxy and F5 router component to work with ingresses
### Not currently planned
- The image build component - a pattern for building Docker images on the cluster as jobs (launch a container that can run a build and then push an image into a registry)
- A version of the Docker registry that stores its metadata in etcd and exposes access to track, watch, and manage images via a Kube compatible API, and authenticates back to the cluster (allows both builds and deployments to iteratively

**Comment by smarterclayton** [comment]
@bgrant0607 this is for reference on our side - we can discuss all this post 1.0

**Comment by bgrant0607** [comment]
Thanks. To the list I'd add:
* L7 - routes #561
* templates / parameterization
* kubectl extensions, such as edit #8263 and secret generation #4822
* any UI elements that can be shared
* misc. patches you're carrying

**Comment by bgrant0607** [comment]
cc @erictune @thockin @davidopp

**Comment by mikedanese** [comment]
- HAProxy config generation framework (and then maybe an haproxy nanny app in that could do direct-to-pod external load balancing).

**Comment by bgrant0607** [comment]
Mentioned in #11368:

kubectl export:
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/export.go
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/exporter.go

Invoking PrepareForCreate was a  good idea, though I think we should do that server-side instead. That's perhaps related to the `/clone` or `/instantiate` idea we've discussed in the past.

**Comment by bprashanth** [comment]
@mikedanese https://github.com/GoogleCloudPlatform/kubernetes/pull/11679

**Comment by smarterclayton** [comment]
I think I want to upstream the "volume" command as well - https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/volume.go

**Comment by bgrant0607** [comment]
Filed #13038 for export.

Would love to see some movement on secret generation #4822, particular since we're about to add config data, also.

**Comment by smarterclayton** [comment]
Hopefully will have some folks free next week.  The people who would
normally upstream those are on vacation.

On Fri, Aug 21, 2015 at 4:01 PM, Brian Grant <notifications@github.com>
wrote:

> Filed #13038 <https://github.com/kubernetes/kubernetes/issues/13038> for
> export.
>
> Would love to see some movement on secret generation #4822
> <https://github.com/kubernetes/kubernetes/issues/4822>, particular since
> we're about to add config data, also.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/10408#issuecomment-133546087>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bgrant0607** [comment]
set-env: https://github.com/kubernetes/kubernetes/pull/13422#issuecomment-136893405

cc @kubernetes/goog-ux

**Comment by smarterclayton** [comment]
The authn/z parts of this are blocked on https://github.com/kubernetes/kubernetes/pull/12209

**Comment by smarterclayton** [comment]
For 1.2 we should discuss what parts of deployments we try to directly upstream (hooks, triggers, custom deployments, extra strategies) and how we manage extension of that object.  Porting routes to ingress and moving the ha proxy and f5 full featured implementation upstream (or simply API compatible) are also top concerns.

**Comment by bgrant0607** [comment]
Another command: env

From #3885: How much of the admin CLI is openshift-specific?
https://github.com/openshift/origin/blob/master/pkg/cmd/admin/admin.go

**Comment by smarterclayton** [comment]
Admin CLI is three parts today:

1. Config related - tools for generating certs, config files, and
validating config - mostly OpenShift specific today, but once we have
config files for Kube the patterns are reasonably applicable and help in
bootstrapping new config
2. Templating router, registry, and ipfailover (keepalived) - these
commands are specific generators to make it easy to parameterize and guide
a user through creating the core components.  I would *prefer* these be
templates, but at the time our security story (secrets) was not good
enough.  We *should* be able to express these only with very limited
templates IMO.  Probably better served by developing template API.
3. Node management, common cluster operations, policy - I think these
should all be in core Kube, since they are admin focused experiential
tools.  #3885 is an example where we may want automatic and manual tools.

Note that everything in admin cli *could* be in the user CLI, but we made a
somewhat arbitrary distinction to split them.  We have two
"create-namespace" commands, one focused on admins for initializing policy
behaviors, and one for end users where they request a namespace (where an
admin may configure the system to deny that).  There was a concern about
polluting the client CLI space with specific or advanced verbs.  That could
be mitigated with "kubectl admin *" or "kubectl cluster *" prefixes.

**Comment by bgrant0607** [comment]
How hard would it be to upstream templates? https://docs.openshift.org/latest/dev_guide/templates.html

**Comment by smarterclayton** [comment]
The code is pretty self contained - questions:

1. Are we ok with the pattern of POST template to server, get back template
or do we want to do the transform on the client in a generator.  Having the
API endpoint makes it easy for multiple clients, having the latter keeps it
self contained on the client
2. Do we need a technical solution for parameterizing numbers or arrays, or
can we start with string substitution and expand in the future
3. Are we ok with a simple "process" command to generate the template (that
you have to pipe to create -f -)
4. Do we want to create the API storage on the server (so you can retrieve
/ view them) or start with just the client side version that the server
transforms

**Comment by bgrant0607** [comment]
cc @kubernetes/rh-dev-experience @kubernetes/rh-ux

**Comment by bgrant0607** [comment]
@smarterclayton Re. templates:

1. Performing substitution in the server is usually preferable, since it works for non-kubectl clients, as you point out, and is generally the direction we want to go #12143. If we could also link the code into kubectl, that might be nice for a --local mode.
2. Parameterizing numbers, probably. Maybe a solution to #17095 would help? Arrays, maybe container commands, though perhaps those use cases could be addressed with another mechanism.
3. Seems like a reasonable starting point, and would also work with apply, delete, etc.
4. Yes, I'd like API storage. It would start in extensions/v1beta1.

**Comment by brendandburns** [comment]
A couple of thoughts:
  * given deployment manager, I'd prefer `{{` to `${`
  * the pipe syntax is a non-starter from my perspective, `oc process <foo> | oc create -f -` is just to complicated/verbose, I'd much rather use the stream transform approach described in #14993
  * we should just start with simple string substitution.  there are lots of other bits upstream from that that will add the necessary validation.

**Comment by bgrant0607** [comment]
@brendandburns 

* If we use the same substitution syntax as Deployment Manager (`{{`), then Deployment Manager could not be composed with the Templates API without escaping and also any differences in behavior may be confusing. 
* I agree that piping (or temporary files) is not ideal long term, but I was looking for an easy way to start, and I disagree that it's a non-starter. It is a straightforward means of composition, and the only means currently for externally built tools. We'd need to make the stream transforms work in (virtually) every command that currently accepts -f/--filename. A search for AddJsonFilenameFlag finds 16 commands that accept --filename.
* I'm fine with starting with strings only (not ints or arrays), if that's what you mean. If you mean #14918, I stated why I prefer the Templates API here: https://github.com/kubernetes/kubernetes/pull/14918#issuecomment-153954995

**Comment by smarterclayton** [comment]
@bparees when you add your proposal ref it here please.

On Nov 17, 2015, at 2:24 AM, Brian Grant <notifications@github.com> wrote:

@brendandburns <https://github.com/brendandburns>

   - If we use the same substitution syntax as Deployment Manager ({{),
   then Deployment Manager could not be composed with the Templates API
   without escaping and also any differences in behavior may be confusing.
   - I agree that piping (or temporary files) is not ideal long term, but I
   was looking for an easy way to start, and I disagree that it's a
   non-starter. It is a straightforward means of composition, and the only
   means currently for externally built tools. We'd need to make the stream
   transforms work in (virtually) every command that currently accepts
   -f/--filename. A search for AddJsonFilenameFlag finds 16 commands that
   accept --filename.
   - I'm fine with starting with strings only (not ints or arrays), if
   that's what you mean. If you mean #14918
   <https://github.com/kubernetes/kubernetes/pull/14918>, I stated why I
   prefer the Templates API here: #14918 (comment)
   <https://github.com/kubernetes/kubernetes/pull/14918#issuecomment-153954995>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/10408#issuecomment-157297768>
.

**Comment by liggitt** [comment]
> 2. Do we need a technical solution for parameterizing numbers or arrays, or can we start with string substitution and expand in the future

transforms of parameter values into primitive values like numbers or booleans, or other encodings (specifically base64 to allow parameter values to be used to construct secrets) are something we were still trying to figure out for OpenShift templates. https://github.com/openshift/origin/issues/3946#issuecomment-135292216 is the most recent conversation we had about it

**Comment by smarterclayton** [comment]
Opened https://github.com/kubernetes/kubernetes/issues/17440 to track auth/authz upstreaming as an ecosystem project as discussed at Kubecon

**Comment by jackgr** [comment]
Perhaps I'm missing something, but why are we talking about enshrining the OpenShift templating and parameterization mechanisms in Kubernetes, when there are existing alternatives, some of which may be better for Kube users?

Wouldn't it be better to let Kube users choose from among the alternatives, see which ones emerge as the solutions that users prefer, and then incorporate those into the code base?

**Comment by smarterclayton** [comment]
OpenShift templates were designed from the beginning as "how would we do simple templatization in a consumable way for users of a Kube platform that leverage all of the core Kube patterns".  The patterns they expose are the same principles we've built Kube around.  

They solve a particular problem - how do I consume, reuse, and expose precanned API object groups to a server, without requiring a full configuration processing tool, but without attempting to solve complex coordination.  They also encode simple input, generation, and feedback to end users without precluding more complex situations.  That has been demonstrated useful to users across a wide range of environments - we've got production applications built using them, and identified gaps and future directions already.

I don't think templates as described are counter to other templating and parameterization mechanisms - they are instead complementary because they are api objects on their own, sit at the sweet spot of value / reward, and don't attempt to solve orchestration and reconciliation (deployment manager), complex templating (deployment manager or any template lange), yet are strictly better than a pure list of api objects (which is scarcely portable).

Discussions about the suitability of the design choices is really something that should be done against the proposal - I think the proposal should be able to stand on its own.

**Comment by jackgr** [comment]
Similar arguments can be made about other available templating solutions. They all have merit. That's why it doesn't make sense to enshrine one of them. At least, not without a shoot out. If we want to drive a decision, instead of waiting for the community to demonstrate a preference through adoption, then we should compile a list of available candidates and a set of evaluation criteria, and then compare them.

**Comment by smarterclayton** [comment]
I think maybe we're talking about different things - the proposal here is
to create an API object that is the framework for simple parameterization.
We can argue about whether that is desirable or not (lots of data
supporting it's the #1 thing that app authors want).  Templating (logic) is
out of scope.  Orchestration (reconciliation) is out of scope.
It's less about templating than about creating a tool that can be composed
and used by other items.

Features that are important are:

1. The objects remain API objects and valid JSON - any Jinja style
transformation is not possible in that form.  Only declarative (change
these fields) are possible, but they don't support complex transformations
easily.  So the approach is the simple variable replacement we use in
environment variables today
2. Objects are introspectable as objects, and only the final pass is
transformative
3. Our other API objects are *easily* used to build real applications
without anything more than simple variable replacement - i.e. name
substitution, port substitution is ok, complex logic is not.

**Comment by bgrant0607** [comment]
@smarterclayton Why is it not possible/desirable to provide template parameters separately from the template when passing it to the substitution pass? I'd like users to be able to specify parameters without necessarily looking at the template.

**Comment by smarterclayton** [comment]
You mean to specify parameters that change things not specified in the
template as parameters?  Or something else?

Why is it not possible/desirable to provide template parameters separately
from the template when passing it to the substitution pass? I'd like users
to be able to specify parameters without necessarily looking at the
template.

**Comment by fabianofranz** [comment]
Also: https://github.com/kubernetes/kubernetes/issues/23402

**Comment by kargakis** [comment]
@smarterclayton what's left here? We definitely miss some `set` commands.

tagging @kubernetes/sig-cli-misc @fabianofranz

**Comment by smarterclayton** [comment]
Yeah set probe languished because we were busy refactoring everything else

**Comment by bgrant0607** [comment]
I think this umbrella issue is mostly obsolete, so closing in favor of more specific ones
